src/wp-admin/install.php

```diff
@@ -299,119 +170,286 @@ function display_setup_form( $error = null ) {
 		$scripts_to_print[] = 'user-profile';
-
+		// Print the branded page header.
 		display_header();
-		// Fill in the data we gathered
+		// Get and remove whitespace from submitted data.
 		$weblog_title = isset( $_POST['weblog_title'] ) ? trim( wp_unslash( $_POST['weblog_title'] ) ) : '';
 		$user_name = isset($_POST['user_name']) ? trim( wp_unslash( $_POST['user_name'] ) ) : '';
 		$admin_password = isset($_POST['admin_password']) ? wp_unslash( $_POST['admin_password'] ) : '';
 		$admin_password_check = isset($_POST['admin_password2']) ? wp_unslash( $_POST['admin_password2'] ) : '';
 		$admin_email  = isset( $_POST['admin_email'] ) ?trim( wp_unslash( $_POST['admin_email'] ) ) : '';
 		$public       = isset( $_POST['blog_public'] ) ? (int) $_POST['blog_public'] : 1;
-
-		// Check email address.
+		// Initialize an error flag.
 		$error = false;
+		// Validate input; toggle error flag if any problems validating.
 		if ( empty( $user_name ) ) {
-			// TODO: poka-yoke
-			display_setup_form( __( 'Please provide a valid username.' ) );
-			$error = true;
+			$error = __( 'Please provide a valid username.' );
 		} elseif ( $user_name != sanitize_user( $user_name, true ) ) {
-			display_setup_form( __( 'The username you provided has invalid characters.' ) );
-			$error = true;
+			$error = __( 'The username you provided has invalid characters.' );
 		} elseif ( $admin_password != $admin_password_check ) {
-			// TODO: poka-yoke
-			display_setup_form( __( 'Your passwords do not match. Please try again.' ) );
-			$error = true;
+			$error = __( 'Your passwords do not match. Please try again.' );
 		} elseif ( empty( $admin_email ) ) {
-			// TODO: poka-yoke
-			display_setup_form( __( 'You must provide an email address.' ) );
-			$error = true;
+			$error = __( 'You must provide an email address.' );
 		} elseif ( ! is_email( $admin_email ) ) {
-			// TODO: poka-yoke
-			display_setup_form( __( 'Sorry, that isn&#8217;t a valid email address. Email addresses look like <code>username@example.com</code>.' ) );
-			$error = true;
+			$error = __( 'Sorry, that isn&#8217;t a valid email address. Email addresses look like <code>username@example.com</code>.' );
 		}
-
+		if ( $error ) {
+			display_setup_form( $error );
+		}
+		/**
+		 * No validation errors? Great. Inform user of success; show username;
+		 * include password note; link to login page.
+		 */
 		if ( $error === false ) {
 			$wpdb->show_errors();
 			$result = wp_install( $weblog_title, $user_name, $admin_email, $public, '', wp_slash( $admin_password ), $loaded_language );
-?>
+			echo '<h1>' . __( 'Installation Complete' ) . '</h1>' . "\n";
+			echo '<p>' . __( 'Your ClassicPress site is ready to go!' ) . '</p>' . "\n";
+			echo '<table class="form-table install-success">' . "\n";
+			echo '<tr>' . "\n";
+			echo '<th>' . __( 'Username' ) . '</th>' . "\n";
+			echo '<td><code>' . esc_html( sanitize_user( $user_name, true ) ) . '</code></td>' . "\n";
```

Comment:
Why do we need to sanitize the user at this point?

Reply:
We've already verified that the user is sanitized.  I've removed this call.